### PR TITLE
[Merged by Bors] - feat : port CategoryTheory.Limits.FilteredColimitCommutesFiniteLimit

### PR DESCRIPTION
**Pull request by kbuzzard**
This was harder than it looked (the proofs are long and broke).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

There are a couple of *TODO*s which can be done before or after merging

### CODE REVIEW

**Comment by semorrison** [comment]
I pushed some minor fixes. Thank you for the heroic effort on this one. :-) I'm sorry I wrote such a long proof back in mathlib3, it must have been annoying waiting for it to check here. :-( At least I left comments?

I agree that your **TODO**s warrant a bit more attention before merging.

**Comment by semorrison** [comment]
The error you got from 

> why doesn't `rw [← types_comp_apply]` work? **TODO** ask on github

which was something like `pattern ?f (?g ?x) was a metavariable, is usually solved by adding arguments to the rw lemma, e.g. in this case `rw [← types_comp_apply (F.map _) (F.map _)]`.

**Comment by kbuzzard** [comment]
> The error you got from
> 
> > why doesn't `rw [← types_comp_apply]` work? **TODO** ask on github
> 
> which was something like `pattern ?f (?g ?x) was a metavariable, is usually solved by adding arguments to the rw lemma, e.g. in this case `rw [← types_comp_apply (F.map _) (F.map _)]`.

`      rw [←types_comp_apply (F.map _) (F.map _) (y j)]` doesn't work either.

**Comment by semorrison** [comment]
See my last push; I just turned the original simp into a rw. In some intermediate state I did have that `rw` above working.

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+      fun {j} {j'} f => by
+      have q :
+        ((curry.obj F).obj j').map (gf f) (F.map _ (y j')) =
+          ((curry.obj F).obj j').map (hf f) (F.map _ (y j)) :=
+        (w f).choose_spec.choose_spec.choose_spec
+      dsimp [curry_obj_obj_obj, curry_obj_obj_map] at q
+      -- porting note: Lean 4 `dsimp` is unfolding `gf` and `hf` :-(
+      -- We fold them back up.
+      -- **TODO** minimise and ask why this is happening
+      change F.map ((𝟙 j', gf f) : (j', k') ⟶ (j', kf f))
```
Comment (kbuzzard) [inline Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:217]: I asked [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.60dsimp.60.20unfolding.20local.20lets/near/352671008) about the random unfolding.

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+variable {J K : Type v} [SmallCategory J] [SmallCategory K]
+
+variable (F : J × K ⥤ Type v)
+
+-- **TODO** naming? placing?
+/-- `(curry.obj F ⋙ lim).obj S` = `limit ((curry.obj F).obj S)` definitionally, so this
+is just a variant of `limit_ext'`. -/
+@[ext] lemma curry_obj_ggg_lim_obj_ext (x y : (curry.obj F ⋙ lim).obj S) (w : ∀ (j : K),
+    limit.π ((curry.obj F).obj S) j x = limit.π ((curry.obj F).obj S) j y) : x = y :=
+  limit_ext' _ x y w
```
Comment (kbuzzard) [inline Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:54]: This fixes an `ext` problem in the file. Any suggestions for names and/or location?

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean]: Great! I agree it's not ideal. If you add a link from here to the zulip discussion I'm happy to merge this part as is.

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:54]: The fact that `ext` won't just do this is really making me sad. :-( It is defeating the design principle that everything boring in the category theory library should be aspirationally be provable `by intros; ext; simp`.

I could just write an `ext!` that isn't a squeamish as the built-in one?

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean (hunk 1 above)
Comment (kbuzzard) [reply Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean]: The link is there.

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean (hunk 2 above)
Comment (ChrisHughes24) [reply Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:54]: This lemma has the wrong name. Can `curry` just be replaced with an arbitrary functor or can this be generalized in some way (I don't think so).

Comment (kbuzzard) [reply Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:51]: I tried removing `curry` when I wrote it and I failed, but I just tried again and it worked fine. What do you think of the current name/statement @ChrisHughes24 ?

Comment (semorrison) [reply Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:51]: Looks good to me!

File: Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+      ∀ {j j'} (f : j ⟶ j'),
+        F.map ((𝟙 j', g j' ≫ gf f) : (j', k j') ⟶ (j', kf f)) (y j') =
+          F.map ((f, g j ≫ hf f) : (j, k j) ⟶ (j', kf f)) (y j) :=
+      fun {j} {j'} f => by
+      have q :
+        ((curry.obj F).obj j').map (gf f) (F.map ((𝟙 j', g j') : (j', k j') ⟶ (j', k')) (y j')) =
+          ((curry.obj F).obj j').map (hf f) (F.map ((f, g j) : (j, k j) ⟶ (j', k')) (y j)) :=
+        (w f).choose_spec.choose_spec.choose_spec
+      rw [curry_obj_obj_map, curry_obj_obj_map] at q
+      -- porting note: Lean 4 `dsimp` unfolds `gf` and `hf` in `q` :-(
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Limits/FilteredColimitCommutesFiniteLimit.lean:213]:
```suggestion
      -- porting note: Lean 4 `dsimp` unfolds `gf` and `hf` in `q` :-(
      -- See discussion at https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/.60dsimp.60.20unfolding.20local.20lets
```

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**